cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java

```diff
@@ -309,6 +311,27 @@ public void testMapreduceHistory() throws Exception {
   }
 
   /**
+   * Tests history of a non existing program
+   */
+  @Test
+  public void testNonExistingProgramHistory() throws Exception {
+    ProgramId program = new ProgramId(TEST_NAMESPACE2, DUMMY_APP_ID, ProgramType.MAPREDUCE, DUMMY_MR_NAME);
+    try {
+      deploy(DummyAppWithTrackingTable.class, Constants.Gateway.API_VERSION_3_TOKEN, TEST_NAMESPACE2);
+      // first run
+      int historyStatus = doPost(getVersionedAPIPath("apps/" + DUMMY_APP_ID + ProgramType.MAPREDUCE + "/NonExisting",
+                                                     Constants.Gateway.API_VERSION_3_TOKEN,
+                                                     TEST_NAMESPACE2)).getStatusLine().getStatusCode();
+      int deleteStatus = doDelete(getVersionedAPIPath("apps/" + DUMMY_APP_ID, Constants.Gateway.API_VERSION_3_TOKEN,
+                                                      TEST_NAMESPACE2)).getStatusLine().getStatusCode();
+      Assert.assertEquals(200, deleteStatus);
```

Comment:
Don't you need to verify the earlier status for not found (404) Look at the code change which terence suggested.